### ISSUE TITLE
use F.relu6 instead of F.relu

### ISSUE THREAD

**Issue by notlober** [KellerJordan/modded-nanogpt#94]
tried relu6, it enables 3.28 in like 1600 iters and reaches 3.24 at end. I had to do gradient accumulation and some less val_tokens but it should mostly stay same, you should definitely try this.

[b79855b1-9b23-472c-a3dc-224b8868a8fd.txt](https://github.com/user-attachments/files/19656893/b79855b1-9b23-472c-a3dc-224b8868a8fd.txt)

**Comment by notlober** [comment]
```
class MLP(nn.Module):
    def __init__(self, dim: int):
        super().__init__()
        hdim = 4 * dim
        self.c_fc = CastedLinear(dim, hdim)
        self.c_clip_min = CastedLinear(dim, hdim)
        self.c_clip_max = CastedLinear(dim, hdim)
        self.c_proj = CastedLinear(hdim, dim)
        self.c_proj.weight.detach().zero_() # zero init suggested by @Grad62304977

    def forward(self, x: Tensor):
        x = torch.min(torch.max(self.c_fc(x), self.c_clip_max(x)), self.c_clip_min(x)).square() # https://arxiv.org/abs/2109.08668v2; ~1-2% better than GELU; suggested by @SKYLINEZ007 and @Grad62304977
        x = self.c_proj(x)
        return x
```

`step:1495/1770 val_loss:3.2778 train_time:1583613ms step_avg:1059.27ms
`
`step:1770/1770 val_loss:3.2129 train_time:1887538ms step_avg:1066.41ms
`

**Comment by notlober** [comment]
![Image](https://github.com/user-attachments/assets/ca309264-7c34-485e-bf10-0f6281548e49)

**Comment by notlober** [comment]
thanks for the feedback, you're correct on most points especially for mlp one, the graph is looking strange because I do evals each 5 iterations in my original logs, but 125 per for your original repo logs so 125 iterations look like a sharp line rather than soft line of 5 per iterations, comparing them directly led to that

but my results are consistent with that exact code I shared and gets better loss (if equally compared, eg. I changed val_tokens), my only modifications were: 

gradient accumulation: 
```
grad_accum_steps = 8 # 1xH100 * 8 = 8xH100
...
    for i in range(grad_accum_steps):
        inputs, targets = next(train_loader)
        model(inputs, targets, get_window_size_blocks(step)).backward()
```

val_tokens and val_seq_len changed from 1m and 256k to 48k and 48k (maybe reason of better(?) convergence is because length generalization since val_seq_len is more close to train_seq_len now):

```
    val_tokens = 48*1024
    train_seq_len = 48*1024
    val_seq_len = 48*1024
    val_loss_every = 5
```

and lastly relu6 modification:
`        x = F.relu6(x).square()`

**Comment by MeetThePatel** [comment]
> We have a limit of ~125M params for the NanoGPT-small track.

@leloykun 

Is this requirement still in effect? The last record (https://github.com/KellerJordan/modded-nanogpt/blob/master/records/020125_RuleTweak/eff63a8c-2f7e-4fc5-97ce-7f600dae0bc7.txt) is showing 275,598,388 trainable params for:
```
trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
```
Am I counting the params wrong?

**Comment by leloykun** [comment]
Hmm, I'm leaning on counting the pre-matmul params so we could encourage the use of crazier sparsity tricks. But iirc, the reason we never switched to using lora style MLPs was because they were (a bit) worse--I'd like to be proven wrong tho!
